x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.ts

```diff
@@ -357,17 +408,29 @@ export const performBulkActionRoute = (
 
                 throwAuthzError(await mlAuthz.validateRuleType(rule.params.type));
 
+                const migratedRule = await legacyMigrate({
+                  rulesClient,
+                  savedObjectsClient,
+                  rule,
+                });
+
+                // This should only be hit if `rule` passed into `legacyMigrate`
+                // is `null` or `rule.id` is null which right now, as typed, should not occur
+                if (migratedRule == null) {
+                  throw new BadRequestError(`An error occurred editing rule with id:${rule.id}`);
+                }
+
                 const editedRule = body[BulkAction.edit].reduce(
                   (acc, action) => applyBulkActionEditToRule(acc, action),
-                  rule
+                  migratedRule
                 );
 
                 const { tags, params: { timelineTitle, timelineId } = {} } = editedRule;
                 const index = 'index' in editedRule.params ? editedRule.params.index : undefined;
 
                 await patchRules({
                   rulesClient,
-                  rule,
+                  rule: editedRule,
```

Comment:
I think `rule` parameter here refers to the original rule, which should be `migratedRule` for this `patchRules` call if I'm not mistaken.

Reply:
On the other hand, perhaps it could work with `rule: editedRule` as well 🤔

Reply:
@yctercero oh yeah, that's a good point! We've overlooked it.

Ok, let me check it separately. I think we will need to get rid of the `patchRules` call here because it doesn't play nice with resetting timeline templates to `None` (details in https://github.com/elastic/kibana/pull/128691), so we'll either have to replace `patchRules` with another function or fix it. Long term we won't need it because we will have a `bulkEdit` method in `RulesClient` (details in https://github.com/elastic/kibana/issues/124715).

No need for any more changes here! 👍

Reply:
@yctercero and everyone else looking here,

In a new [PR](https://github.com/elastic/kibana/pull/129491) with the fix of using `patchRules` in bulk editing, I replaced the call to `patchRules` here with a new specialized function `editRule`.

```ts
export const editRule = async (params: EditRuleParams): Promise<RuleAlertType> => {
  const { rulesClient, rule, edit } = params;
  const isPrebuiltRule = rule.params.immutable;
  const isCustomRule = !rule.params.immutable;

  if (isPrebuiltRule) {
    throw new EditRuleError('Elastic rule can`t be edited', 400);
  }

  const editedRule = applyChanges(rule, edit);

  // If the rule wasn't changed by the `edit` function, we don't need to proceed with the update.
  if (!isRuleChanged(rule, editedRule)) {
    return rule;
  }

  // We need to increment the rule's version if it is a custom rule. If the rule is an Elastic
  // prebuilt rule, we don't want to touch its version - it's managed by the rule authors.
  // This check is left here explicitly because we're planning to allow editing for prebuilt rules,
  // and the check for isPrebuiltRule above might be removed.
  if (isCustomRule) {
    editedRule.params.version = editedRule.params.version + 1;
  }

  const updateData = createUpdateData(rule, editedRule);
  await rulesClient.update({
    id: rule.id,
    data: updateData,
  });

  // It would be great to return the updated rule returned from the RulesClient.update() call.
  // Note that there's a type mismatch between RuleAlertType and the update method result.
  return editedRule;
};
```

So here it returns the rule that we edited in memory, rather than the updated rule returned from the RulesClient. It's because the latter one has a different type than the one we need. I'm going to open a ticket for fixing it in the RulesClient.

Reply:
Thanks for the update here @banderror !